driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java

```diff
@@ -247,46 +246,43 @@ private boolean handleReplicaSetMemberChanged(final ServerDescription newDescrip
         if (newDescription.getCanonicalAddress() != null
                 && !newDescription.getAddress().equals(new ServerAddress(newDescription.getCanonicalAddress()))
                 && !newDescription.isPrimary()) {
-            if (LOGGER.isInfoEnabled()) {
-                LOGGER.info(format("Canonical address %s does not match server address.  Removing %s from client view of cluster",
-                                   newDescription.getCanonicalAddress(), newDescription.getAddress()));
-            }
+            LOGGER.info(format("Canonical address %s does not match server address.  Removing %s from client view of cluster",
+                    newDescription.getCanonicalAddress(), newDescription.getAddress()));
             removeServer(newDescription.getAddress());
             return true;
         }
 
         if (newDescription.isPrimary()) {
             ObjectId electionId = newDescription.getElectionId();
             Integer setVersion = newDescription.getSetVersion();
-            if (setVersion != null && electionId != null) {
-                if (isStalePrimary(newDescription)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Invalidating potential primary %s whose (set version, election id) tuple of (%d, %s) "
-                                + "is less than one already seen of (%d, %s)",
-                                newDescription.getAddress(),
-                                setVersion, electionId,
-                                maxSetVersion, maxElectionId));
-                    }
-                    addressToServerTupleMap.get(newDescription.getAddress()).server.resetToConnecting();
+            if (newDescription.getMaxWireVersion() >= SIX_DOT_ZERO_WIRE_VERSION) {
+                if (nullSafeCompareTo(electionId, maxElectionId) > 0
+                        || (nullSafeCompareTo(electionId, maxElectionId) == 0 && nullSafeCompareTo(setVersion, maxSetVersion) >= 0)) {
+                    LOGGER.info(format("Setting max election id to %s and max set version to %d from replica set primary %s",
+                            newDescription.getElectionId(), newDescription.getSetVersion(), newDescription.getAddress()));
+                    maxElectionId = newDescription.getElectionId();
+                    maxSetVersion = newDescription.getSetVersion();
+                } else {
+                    invalidatePotentialPrimary(newDescription);
                     return false;
                 }
-
-                if (!electionId.equals(maxElectionId)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Setting max election id to %s from replica set primary %s", electionId,
-                                newDescription.getAddress()));
+            } else {
+                if (setVersion != null && electionId != null) {
+                    if (nullSafeCompareTo(maxSetVersion, setVersion) <= 0
```

Comment:
I removed the null checks for maxSetVersion and maxElectionId from the spec because `nullSafeCompareTo` makes them redundant